File: 2021/Day 03/aoc03_2.cpp
#include <iostream>
#include <fstream>
#include <vector>
using namespace std;

int main() {
    ifstream infile;
    infile.open("aoc03_input.txt");

    vector<string> l {};
    
    string d;
    while(infile >> d) {
        l.push_back(d);
    }
    
    int n = l[0].size();

    vector<string> s = l;
    int a = 0;
    for(int i = 0; i < n; i++) {
        int ones = 0;
        int zeros = 0;

        for(string j : s) {
            if(j[i] == '0')
                zeros++;
            else
                ones++;
        }

        a <<= 1;
        if(ones >= zeros)
            a |= 1;
        
        vector<string> ns {};
        for(string j : s) {
            if((int) (j[i] - 0x30) == (int) (a & 1)) {
                ns.push_back(j);
            }
        }
        
        s = ns;
    }

    s = l;
    int b = 0;
    for(int i = 0; i < n; i++) {
        int ones = 0;
        int zeros = 0;

        if(s.size() == 1) {
            b <<= 1;
            b |= s[0][i] - 0x30;
            continue;
        }

        for(string j : s) {
            if(j[i] == '0')
                zeros++;
            else
                ones++;
        }

        b <<= 1;
        if(ones < zeros)
            b |= 1;
        
        vector<string> ns {};
        for(string j : s) {
            if((int) (j[i] - 0x30) == (int) (b & 1)) {
                ns.push_back(j);
            }
        }
        
        s = ns;
    }

    cout << a * b;

    return 0;
}

File: 2021/Day 03/aoc03_2.py
file = open("aoc03_input.txt")
l = file.read().split("\n")
file.close()

s = set(l)
a = ""
for i in range(len(l[0])):
    ones = 0
    zeros = 0
    for j in s:
        if j[0] == "0":
            zeros += 1
        else:
            ones += 1

    if ones >= zeros:
        a += "1"
    else:
        a += "0"
    
    ns = set()
    for j in s:
        if j[0] == a[-1]:
            ns.add(j[1:])
    
    s = ns

a = int(a, 2)

s = set(l)
b = ""
for i in range(len(l[0])):
    if len(s) == 1:
        break

    ones = 0
    zeros = 0
    for j in s:
        if j[0] == "0":
            zeros += 1
        else:
            ones += 1

    if ones < zeros:
        b += "1"
    else:
        b += "0"
    
    ns = set()
    for j in s:
        if j[0] == b[-1]:
            ns.add(j[1:])
    
    s = ns

b = int(b + list(s)[0], 2)

print(a, b)

print(a * b)